### PR TITLE
ENH: Generate the docstrings of umath into a separate C header

### PR DESCRIPTION
**Pull request by seiko2plus**
To allow access to the docstring of ufunc via identifier #DOC_PLACE_OF_UFUNC, e.g. DOC_NUMPY_CORE_UMATH_ABSOLUTE.

  This patch is part of a series of steps aimed at replacing template sources
  and the umath python generator with C++ meta-programming.

It was part of #19753

### CODE REVIEW

File: numpy/core/setup.py [hunk 1]
```diff
@@ -965,6 +965,19 @@ def generate_umath_c(ext, build_dir):
...
+    def generate_umath_doc_header(ext, build_dir):
+        target = join(build_dir, header_dir, '_umath_doc_generated.h')
+        dir = os.path.dirname(target)
+        if not os.path.exists(dir):
+            os.makedirs(dir)
+
+        generate_umath_doc_py = join(codegen_dir, 'generate_umath_doc.py')
+        if newer(generate_umath_doc_py, target):
+            n = dot_join(config.name, 'generate_umath_doc')
+            generate_umath_doc = npy_load_module(
```
Comment (charris) [inline numpy/core/setup.py:977]: This file seems to be the only place that uses `npy_load_module`, and that function uses tools that will be deprecated in Python 3.12. I wonder if we should use the alternative `numpy.distutils.misc_util.exec_mod_from_location` instead?

File: numpy/core/code_generators/generate_umath_doc.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    # Split the docstring because some compilers (like MS) do not like big
+    # string literal in C code. We split at endlines because textwrap.wrap
+    # do not play well with \n
+    docstring = '\\n\"\"'.join(docstring.split(r"\n"))
+    return docstring
+
+def write_code(target):
+    with open(target, 'w') as fid:
+        fid.write(
+            "#ifndef _NPY_UMATH_DOCSTRING_H_\n"
```
Comment (charris) [inline numpy/core/code_generators/generate_umath_doc.py:23]: This is OK, but I have slowly been moving header guards to use the full path. Not sure it is worth the effort for a generated file.

File: numpy/core/code_generators/generate_umath.py [hunk 3]
```diff
@@ -17,6 +13,15 @@
 MinusInfinity = 'PyFloat_FromDouble(-NPY_INFINITY)'
 ReorderableNone = "(Py_INCREF(Py_None), Py_None)"
 
+class docstrings:
+    def get(place):
+        """
+        Returns the C #definition name of docstring according
+        to ufunc place. C #definitions are generated by generate_umath_doc.py
+        in a separated C header.
```
Comment (charris) [inline numpy/core/code_generators/generate_umath.py:21]: separated <- separate

**Comment by charris** [comment]
LGTM apart from a few nits.  We may be moving to building with meson in the next year or so, there might be things we could do to smooth the way.  @rgommers Thoughts?

**Comment by rgommers** [comment]
> We may be moving to building with meson in the next year or so, there might be things we could do to smooth the way. @rgommers Thoughts?

I don't think we have to worry about it for this kind of codegen change. This all looks good and is not difficult to support/reimplement in Meson.

File: numpy/core/setup.py (hunk 1 above)
Comment (seiko2plus) [reply numpy/core/setup.py]: I have replaced the both cases of `npy_load_module` to `exec_mod_from_location`.

File: numpy/core/code_generators/generate_umath_doc.py (hunk 2 above)
Comment (seiko2plus) [reply numpy/core/code_generators/generate_umath_doc.py]: better to respect the new approach. I have replace it to the full path.

**Comment by charris** [comment]
Thanks Sayed.

**Pull request merged**